File: frontend/src/api/customers.ts
import { http } from "./api";
import type { PaginatedResponse, Customer } from "./types";

export interface CustomerCreatePayload {
  full_name: string;
  email: string;
  phone?: string;
  address?: string;
}

export const fetchCustomers = async (page = 1, limit = 20, search = "") => {
  const response = await http.get<PaginatedResponse<Customer>>("/customers", {
    params: { page, limit, search },
  });
  return response.data;
};

export const createCustomer = async (payload: CustomerCreatePayload) => {
  const response = await http.post<Customer>("/customers", payload);
  return response.data;
};

export const updateCustomer = async (id: number, payload: Partial<CustomerCreatePayload>) => {
  const response = await http.put<Customer>(`/customers/${id}`, payload);
  return response.data;
};

export const deleteCustomer = async (id: number) => {
  const response = await http.delete<Customer>(`/customers/${id}`);
  return response.data;
};

File: backend/app/routers/customers.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_session
from app import crud, schemas

router = APIRouter(
    prefix="/customers",
    tags=["Customers"]
)


@router.get("", response_model=dict)
async def list_customers(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    search: str | None = Query(None),
    session: AsyncSession = Depends(get_session),
):
    items, total = await crud.list_customers(
        session,
        page,
        limit,
        search,
    )

    return {
        "data": [
            schemas.CustomerRead
            .model_validate(item)
            .model_dump()
            for item in items
        ],
        "pagination": {
            "page": page,
            "limit": limit,
            "total": total,
        },
    }


@router.get(
    "/{customer_id}",
    response_model=schemas.CustomerRead
)
async def get_customer(
    customer_id: int,
    session: AsyncSession = Depends(get_session),
):
    return await crud.get_customer(
        session,
        customer_id,
    )


@router.post(
    "",
    response_model=schemas.CustomerRead,
    status_code=201
)
async def create_customer(
    payload: schemas.CustomerCreate,
    session: AsyncSession = Depends(get_session),
):
    return await crud.create_customer(
        session,
        payload,
    )


@router.put(
    "/{customer_id}",
    response_model=schemas.CustomerRead
)
async def update_customer(
    customer_id: int,
    payload: schemas.CustomerUpdate,
    session: AsyncSession = Depends(get_session),
):
    return await crud.update_customer(
        session,
        customer_id,
        payload,
    )


@router.delete(
    "/{customer_id}",
    response_model=schemas.CustomerRead
)
async def delete_customer(
    customer_id: int,
    session: AsyncSession = Depends(get_session),
):
    return await crud.delete_customer(
        session,
        customer_id,
    )